File: app/models/event.py
from datetime import datetime
from app import db

class Event(db.Model):
    __tablename__ = 'events'
    
    id = db.Column(db.Integer, primary_key=True)
    title = db.Column(db.String(200), nullable=False)
    description = db.Column(db.Text)
    
    # Date et heure
    start_date = db.Column(db.DateTime, nullable=False)
    end_date = db.Column(db.DateTime)
    
    # Localisation
    location = db.Column(db.String(200))
    address = db.Column(db.Text)
    
    # Type et visibilité
    event_type = db.Column(db.String(50), default='General')  # 'General', 'Department', 'Family'
    target_department_id = db.Column(db.Integer, db.ForeignKey('departments.id'))
    target_family_id = db.Column(db.Integer, db.ForeignKey('families_impact.id'))
    
    # Capacité et inscription
    max_participants = db.Column(db.Integer)
    registration_required = db.Column(db.Boolean, default=False)
    registration_deadline = db.Column(db.DateTime)
    
    # Organisateur
    organizer_id = db.Column(db.Integer, db.ForeignKey('users.id'))
    
    # Statut
    is_active = db.Column(db.Boolean, default=True)
    is_cancelled = db.Column(db.Boolean, default=False)
    
    created_at = db.Column(db.DateTime, default=datetime.utcnow)
    updated_at = db.Column(db.DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)
    
    # Relations
    organizer = db.relationship('User', backref='organized_events')
    target_department = db.relationship('Department', backref='events')
    target_family = db.relationship('FamilyImpact', backref='events')
    participants = db.relationship('EventParticipant', back_populates='event', cascade='all, delete-orphan')
    
    @property
    def current_participants_count(self):
        return len([p for p in self.participants if p.status == 'Confirmed'])
    
    @property
    def available_spots(self):
        if not self.max_participants:
            return None
        return max(0, self.max_participants - self.current_participants_count)
    
    @property
    def is_full(self):
        if not self.max_participants:
            return False
        return self.current_participants_count >= self.max_participants
    
    def __repr__(self):
        return f'<Event {self.title}>'

class EventParticipant(db.Model):
    __tablename__ = 'event_participants'
    
    id = db.Column(db.Integer, primary_key=True)
    user_id = db.Column(db.Integer, db.ForeignKey('users.id'), nullable=False)
    event_id = db.Column(db.Integer, db.ForeignKey('events.id'), nullable=False)
    registered_at = db.Column(db.DateTime, default=datetime.utcnow)
    status = db.Column(db.String(20), default='Confirmed')  # 'Confirmed', 'Cancelled', 'Waiting'
    
    # Relations
    user = db.relationship('User', back_populates='event_participations')
    event = db.relationship('Event', back_populates='participants')
    
    __table_args__ = (db.UniqueConstraint('user_id', 'event_id', name='unique_user_event'),)

File: impact-centre-mobile/backend/models/Event.js
import mongoose from 'mongoose';

const eventSchema = new mongoose.Schema({
  title: {
    type: String,
    required: true,
    trim: true
  },
  description: {
    type: String,
    trim: true
  },
  startDate: {
    type: Date,
    required: true
  },
  endDate: {
    type: Date
  },
  location: {
    type: String,
    trim: true
  },
  address: {
    type: String,
    trim: true
  },
  eventType: {
    type: String,
    enum: ['General', 'Department', 'Family'],
    default: 'General'
  },
  targetDepartmentId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Department'
  },
  targetFamilyId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'FamilyImpact'
  },
  maxParticipants: {
    type: Number
  },
  registrationRequired: {
    type: Boolean,
    default: false
  },
  registrationDeadline: {
    type: Date
  },
  organizerId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true
  },
  isActive: {
    type: Boolean,
    default: true
  },
  isCancelled: {
    type: Boolean,
    default: false
  },
  participants: [{
    userId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User',
      required: true
    },
    registeredAt: {
      type: Date,
      default: Date.now
    },
    status: {
      type: String,
      enum: ['Confirmed', 'Cancelled', 'Waiting'],
      default: 'Confirmed'
    }
  }],
  image: {
    type: String
  },
  color: {
    type: String,
    default: '#28a745'
  }
}, {
  timestamps: true
});

// Index pour recherche et tri
eventSchema.index({ startDate: 1 });
eventSchema.index({ title: 'text', description: 'text' });
eventSchema.index({ eventType: 1, targetDepartmentId: 1 });

// Virtuals
eventSchema.virtual('currentParticipantsCount').get(function() {
  return this.participants.filter(p => p.status === 'Confirmed').length;
});

eventSchema.virtual('availableSpots').get(function() {
  if (!this.maxParticipants) return null;
  return Math.max(0, this.maxParticipants - this.currentParticipantsCount);
});

eventSchema.virtual('isFull').get(function() {
  if (!this.maxParticipants) return false;
  return this.currentParticipantsCount >= this.maxParticipants;
});

// Méthodes
eventSchema.methods.addParticipant = function(userId) {
  const existingParticipant = this.participants.find(
    p => p.userId.toString() === userId.toString()
  );
  
  if (existingParticipant) {
    existingParticipant.status = 'Confirmed';
    return this;
  }
  
  const status = this.isFull ? 'Waiting' : 'Confirmed';
  this.participants.push({ userId, status });
  return this;
};

eventSchema.methods.removeParticipant = function(userId) {
  const participant = this.participants.find(
    p => p.userId.toString() === userId.toString()
  );
  
  if (participant) {
    participant.status = 'Cancelled';
  }
  
  return this;
};

export default mongoose.models.Event || mongoose.model('Event', eventSchema);